controllers/apim/ratelimitpolicy_controller.go

```diff
@@ -85,15 +100,343 @@ func (r *RateLimitPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl
 		}
 	}
 
-	specResult, specErr := r.reconcileSpec(ctx, &rlp)
-	logger.Info("spec reconcile done", "result", specResult, "error", specErr)
-	if specErr == nil && specResult.Requeue {
-		logger.Info("Reconciling not finished. Requeueing.")
-		return specResult, nil
+	if err := r.reconcileLimits(ctx, &rlp); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, networkingRef := range rlp.Spec.NetworkingRef {
+		switch networkingRef.Type {
+		case apimv1alpha1.NetworkingRefType_HR:
+			logger.Info("HTTPRoute is not implemented yet") // TODO(rahulanand16nov)
+			return ctrl.Result{}, nil
+		case apimv1alpha1.NetworkingRefType_VS:
+			vsNamespacedName := client.ObjectKey{
+				Namespace: rlp.Namespace,
+				Name:      networkingRef.Name,
+			}
+
+			var vs istio.VirtualService
+			if err := r.Client().Get(ctx, vsNamespacedName, &vs); err != nil {
+				if apierrors.IsNotFound(err) {
+					logger.Info("no virtualservice found", "lookup name", vsNamespacedName.String())
+					return ctrl.Result{}, nil
+				}
+				logger.Error(err, "failed to get VirutalService")
+				return ctrl.Result{}, err
+			}
+
+			if err := r.reconcileWithVirtualService(ctx, &vs, &rlp); err != nil {
+				logger.Error(err, "failed to reconcile with virtualservice")
+				return ctrl.Result{}, err
+			}
+		}
 	}
 	return ctrl.Result{}, nil
 }
 
+func (r *RateLimitPolicyReconciler) reconcileWithVirtualService(ctx context.Context, vs *istio.VirtualService, rlp *apimv1alpha1.RateLimitPolicy) error {
+	logger := r.Logger()
+
+	_, ok := vs.Annotations[VSAnnotationRateLimitPolicy]
+	if !ok {
+		rlpNamespacedName := client.ObjectKeyFromObject(rlp)
+		vs.Annotations[VSAnnotationRateLimitPolicy] = rlpNamespacedName.String()
+		if err := r.Client().Update(ctx, vs); err != nil {
+			logger.Error(err, "failed to add ratelimitpolicy annotation to virtualservice")
+			return err
+		}
+		logger.V(1).Info("successfully added ratelimitpolicy annotation to virtualservice")
+	}
+
+	// TODO(rahulanand16nov): store context of virtualservice in RLP's status block and manage envoy patches.
+	if err := r.reconcileRateLimitFilters(ctx, vs, rlp); err != nil {
```

Comment:
envoyfilters reconciliation in a method called "reconcileWithVirtualService` feels weird

Reply:
The idea is we have two networking sources, namely, VirtualService and HTTPRoute. So, when VS changes, use `reconcileWithVirtualService` and so forth for HTTPRoute. Reconciling of envoyfilters is required since they rely on gateway names from networking sources.

Reply:
Yes, it feels a bit weird to me as well. I'll change this as well :thinking: